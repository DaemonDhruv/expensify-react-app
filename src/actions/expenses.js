import uuid from 'uuid';
import database from '../firebase/firebase';

// component calls acction generator
// action generator returns the ojbect
// component dispatches object
// redux store changes

// Action generators
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense,
    
});


// This is asyncronous action. It returns a function and then this function is passed to the reducer via the fake dispatch call from component.
// The reducer then runs this function where in we are first pushing the data to the firebase database and on a successful write, call the real 
// add addExpense action which will return the action object consisting of the component-passed data. We then give this action object to the real 
// dispatch call. This will in turn call the reducer and the state will get updated in the redux store.
export const startAddExpense = (expenseData = {}) => {
    // Setup the destructured defaults
    const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = expenseData;
    const expense = {description, note, amount, createdAt };

    // Redux calls this thunk function with a dispatch method
    return (dispatch, getState) => {
        const userId = getState().auth.userId;
        // Push to firebase database
        // returning the bellow will allow us do promise chaining for testing purposes
        return database.ref(`users/${userId}/expenses`)
            .push(expense)
            .then((ref) => {
                // Only after successful write to the DB we want to flush same data into our redux store
                dispatch(addExpense({
                    id: ref.key, // firebase database will create a unique key, under which it will save the expense object. We want to retrive that unique key as our id.
                    ...expense
                }))
            })
            .catch((e) => {
                console.log('Error writing addingExpense to DB: ' + e);
                // Then maybe show this on a modal pop up.
            })
    }
}


// Removing expense by id. Passing Id is compulsory
export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = (id) => {

    return (dispatch, getState) => {
        const userId = getState().auth.userId;
        return database.ref(`users/${userId}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense(id));
            })
            .catch((e) => {
                console.log('Error writing addingExpense to DB: ' + e);
                // Then maybe show this on a modal pop up.
            })
    }
}

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {

    return (dispatch, getState) => {
        const userId = getState().auth.userId;
        return database.ref(`users/${userId}/expenses/${id}`)
            .update(updates)
            .then((ref) => {
                // Only after successful write to the DB we want to flush same data into our redux store
                dispatch(editExpense(id, updates))
            })
            .catch((e) => {
                console.log('Error writing addingExpense to DB: ' + e);
                // Then maybe show this on a modal pop up.
            })
    }
}

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    const expenses = [];
    return (dispatch, getState) => {
        const userId = getState().auth.userId;
        return database.ref(`users/${userId}/expenses`)
            .once('value')
            .then((snapshot) => {
                
                // Parsing the snapshot
                snapshot.forEach(childSnapshot => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(expenses));
            })
            .catch((e) => {
                console.log('Error fetching expenses from DB: ' + e);
                // Then maybe show this on a modal pop up.
            })
    }
}