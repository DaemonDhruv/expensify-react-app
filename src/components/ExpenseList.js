import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

// exporting it separately for testing purposes
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ?
                <p>Expense list is empty</p>
                :
                (
                    props.expenses.map((expense, index) =>
                        <ExpenseListItem
                            key={expense.id}
                            count={index + 1}
                            {...expense}
                        />)
                )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);