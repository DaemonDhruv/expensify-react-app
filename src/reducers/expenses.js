/**
 * Expenses Reducer
 * 
 * @author Dhruv N. Bindoria.
 * @since  1.0.0
 */

/* 
    Array of expense items.
    Each item is an expense object having following properties:
        {
            description,
            note,
            amount,
            createdAt
        }
    Hence, state = [ {d, n, a, c}, {d, n, a, c}, .. ]
*/
const expensesReducerDefaultState = [];

/**
 * Reducer logic for managing our expenses
 *
 * @since      1.0.0
 * 
 * @alias    expensesReducer
 * @memberof reducers
 *
 * @param {Object[]}    state
 * @param {Object}      action   
 *
 * @return {Object[]}   Array of expense items
 */

 // We are not using 'break' here because we are returning
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};