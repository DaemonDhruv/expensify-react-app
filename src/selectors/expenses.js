/**
 * Filtered expense Items
 * 
 * @author Dhruv N. Bindoria.
 * @since  1.0.0
 */

import moment from 'moment';

/**
 * Filters and sorts the expenses.
 *
 * @since      1.0.0
 * 
 * @alias    getVisibleExpenses
 * @memberof selectos
 *
 * @param {Object[]}    expenses    -   Array of all expense objects (state.expenses)
 * @param {Object}      filters     -   Destructured state.filter object.
 *
 * @return {Object[]}   Array of expense items
 */
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter(expense => {
        /* 
            If filters are not set then we want to display all items.
            If it is set then we want to apply filter specific condition.
            https://momentjs.com/docs/#/query/is-same-or-before/ 
        */ 
        const createdAtMoment = moment(expense.createdAt); //  Text to moment object
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'): true;
        const textMatch = text.length == 0 || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        // Sort by date, descending: most recent one, first.
        if(sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1; 
        }
        // Sort by amount, descending: higher amount, first.
        else if(sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
}