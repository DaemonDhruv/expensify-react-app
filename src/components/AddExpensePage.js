import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        // Redirect programatically after successful submit-dispatch to dashboard
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                Add you expense here
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

// By doing this, we are separating redux-react even apart
// And making it more test friendly
// Now the react component don't have to explicitly use 
// redux methods. As all data and methods will be passed 
// in as normal props

// Why we separated concerns for testing purposes?
// Becuase dispatch() need addExpense() and passing that would
// be difficult and out of our scope here. We only are concerned
// with the working of onSubmit. Whether it is getting called
// with proper values or not.
const matchDispatchToProps = (dispatch) => ({
        // This is going be passed on in the props
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, matchDispatchToProps)(AddExpensePage);