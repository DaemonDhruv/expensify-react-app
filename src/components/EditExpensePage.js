import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from '../components/ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        console.log('updated', expense);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onRemove}
                >
                    Remove
                 </button>
            </div>
        );
    }
}

// Here we are searching the id we got from the URL from the router, with the id in the 
// expenses list from the state in the redux store
// And we are extract that particular expense and passing it in the props to our 
// EditExpensePage component
const mapStateToProps = (state, props /* this is the props we got from the router when it called the EditExpensePage component */) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

// mapDispatchToProps = (dispatch, [ownProps]) => ({})
const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);