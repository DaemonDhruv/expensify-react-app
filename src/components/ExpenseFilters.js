import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export class ExpenseFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({
            calendarFocused
        }))
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        //There is a weird bug here... dispatch is so fast.. it
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }
    render() {
        return (
            <div>
                {/* Search by Description */}
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />

                {/* Sort by Date/Amount - Always descending*/}
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>


                {/* Show by Date Range */}
                {/* Learn more at: https://github.com/airbnb/react-dates#daterangepicker */}
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
};


const mapStateToProps = (state) => ({
    filters: state.filters
}); // Along with filters, dispatch() method also gets passed to ExpenseFilters component

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});


// connect() is working as a middleware between parent component and child component. 
// connect() is injecting data from in between to the child component in it's props
// connect() is also providing methods in the props which will take away data and put into the store

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilters);