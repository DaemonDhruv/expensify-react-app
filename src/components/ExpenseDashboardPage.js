import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseFilters from './ExpenseFilters';

const ExpenseDashboardPage = () => (
    <div>
        This is expense dashboard
        <ExpenseFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;