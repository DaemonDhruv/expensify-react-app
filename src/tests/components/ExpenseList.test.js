import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render Expense list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
})

test('should render Expense list if list is empty', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
})