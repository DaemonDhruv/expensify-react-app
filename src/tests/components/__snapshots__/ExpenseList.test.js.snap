// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense list if list is empty 1`] = `
<div>
  <p>
    Expense list is empty
  </p>
</div>
`;

exports[`should render Expense list with expenses 1`] = `
<div>
  <Connect(ExpenseListItem)
    amount={195}
    count={1}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={10900}
    count={2}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={4500}
    count={3}
    createdAt={345600000}
    description="Credit"
    id="3"
    key="3"
    note=""
  />
</div>
`;
